File: solutions/cpp/Rebuild_BST_preorder_better.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <limits>
#include <memory>
#include <vector>

#include "./BST_prototype.h"

using std::cout;
using std::endl;
using std::numeric_limits;
using std::unique_ptr;
using std::vector;

BSTNode<int>* RebuildBSTFromPreorderHelper(const vector<int>& preorder, int min,
                                           int max, size_t* idx);

// @include
BSTNode<int>* RebuildBSTFromPreorder(const vector<int>& preorder) {
  size_t idx = 0;
  return RebuildBSTFromPreorderHelper(preorder, numeric_limits<int>::min(),
                                      numeric_limits<int>::max(), &idx);
}

BSTNode<int>* RebuildBSTFromPreorderHelper(
    const vector<int>& preorder, int min, int max, size_t* idx) {
  if (*idx == preorder.size()) {
    return nullptr;
  }

  int curr = preorder[*idx];
  if (curr < min || curr > max) {
    return nullptr;
  }

  ++*idx;
  return new BSTNode<int>{
      curr, unique_ptr<BSTNode<int>>(
                RebuildBSTFromPreorderHelper(preorder, min, curr, idx)),
      unique_ptr<BSTNode<int>>(
          RebuildBSTFromPreorderHelper(preorder, curr, max, idx))};
}
// @exclude

template <typename T>
void CheckAns(const unique_ptr<BSTNode<T>>& n, const T& pre) {
  if (n) {
    CheckAns(n->left, pre);
    assert(pre <= n->data);
    cout << n->data << endl;
    CheckAns(n->right, n->data);
  }
}

int main(int argc, char* argv[]) {
  //      3
  //    2   5
  //  1    4  6
  // should output 1, 2, 3, 4, 5, 6
  // preorder [3, 2, 1, 5, 4, 6]
  vector<int> preorder;
  preorder.emplace_back(3);
  preorder.emplace_back(2);
  preorder.emplace_back(1);
  preorder.emplace_back(5);
  preorder.emplace_back(4);
  preorder.emplace_back(6);
  unique_ptr<BSTNode<int>> root(RebuildBSTFromPreorder(preorder));
  CheckAns<int>(root, numeric_limits<int>::min());
  return 0;
}

File: solutions/cpp/java/src/main/java/com/epi/RebuildBSTPreorderBetter.java
package com.epi;

import com.epi.BinarySearchTreePrototypeTemplate.BSTNode;

import java.util.ArrayList;
import java.util.List;

public class RebuildBSTPreorderBetter {
  // @include
  private static Integer idx;

  public static BSTNode<Integer> rebuildBSTFromPreorder(
      List<Integer> preorderSequence) {
    idx = 0;
    return rebuildBSFromPreorderHelper(preorderSequence,
        Integer.MIN_VALUE, Integer.MAX_VALUE);
  }

  private static BSTNode<Integer> rebuildBSFromPreorderHelper(
      List<Integer> preorderSequence, Integer min, Integer max) {
    if (idx == preorderSequence.size()) {
      return null;
    }

    Integer curr = preorderSequence.get(idx);
    if (curr < min || curr > max) {
      return null;
    }
    ++idx;
    return new BSTNode<>(
        curr, rebuildBSFromPreorderHelper(preorderSequence, min, curr),
        rebuildBSFromPreorderHelper(preorderSequence, curr, max));
  }
  // @exclude

  private static void checkAns(BSTNode<Integer> n, Integer pre) {
    if (n != null) {
      checkAns(n.getLeft(), pre);
      assert (pre <= n.getData());
      System.out.println(n.getData());
      checkAns(n.getRight(), n.getData());
    }
  }

  public static void main(String[] args) {
    // 3
    // 2 5
    // 1 4 6
    // should output 1, 2, 3, 4, 5, 6
    // preorder [3, 2, 1, 5, 4, 6]
    List<Integer> preorder = new ArrayList<>();
    preorder.add(3);
    preorder.add(2);
    preorder.add(1);
    preorder.add(5);
    preorder.add(4);
    preorder.add(6);
    BSTNode<Integer> tree = rebuildBSTFromPreorder(preorder);
    checkAns(tree, Integer.MIN_VALUE);
  }
}
